File: crewplay-platform/scripts/deploy-sms-vps.mjs
/**
 * Deploy sms-proxy to VPS via SSH/SFTP (Node.js + ssh2)
 */
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";
import { Client } from "ssh2";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

function parseArgs() {
  const args = process.argv.slice(2);
  const opts = {
    host: "103.144.33.38",
    user: "root",
    password: "",
    domain: "sms.crewplay.tw",
    smsDir: path.join(__dirname, "../sms-proxy"),
  };
  for (let i = 0; i < args.length; i += 2) {
    const k = args[i]?.replace(/^--/, "");
    const v = args[i + 1];
    if (k && v) opts[k === "sms-dir" ? "smsDir" : k] = v;
  }
  return opts;
}

function exec(conn, cmd, timeout = 900000) {
  return new Promise((resolve, reject) => {
    conn.exec(cmd, { pty: true }, (err, stream) => {
      if (err) return reject(err);
      let out = "";
      stream
        .on("data", (d) => {
          const s = d.toString();
          out += s;
          process.stdout.write(s);
        })
        .stderr.on("data", (d) => process.stderr.write(d.toString()));
      stream.on("close", (code) => {
        if (code === 0) resolve(out);
        else reject(new Error(`Command failed (${code}): ${cmd}`));
      });
    });
    setTimeout(() => reject(new Error("Timeout")), timeout);
  });
}

function put(sftp, local, remote) {
  return new Promise((resolve, reject) => {
    sftp.fastPut(local, remote, (err) => (err ? reject(err) : resolve()));
  });
}

function mkdir(sftp, remote) {
  return new Promise((resolve) => {
    sftp.mkdir(remote, () => resolve());
  });
}

async function main() {
  const opts = parseArgs();
  if (!opts.password) {
    console.error("Missing --password");
    process.exit(1);
  }

  const appDir = "/opt/crewplay-sms-proxy";
  const files = [
    ["server.mjs", "server.mjs"],
    [".env", ".env"],
    ["deploy/crewplay-sms-proxy.service", "deploy/crewplay-sms-proxy.service"],
    ["deploy/nginx-sms.conf.example", "deploy/nginx-sms.conf.example"],
    ["deploy/install-remote.sh", "deploy/install-remote.sh"],
  ];

  const conn = new Client();
  await new Promise((resolve, reject) => {
    conn
      .on("ready", resolve)
      .on("error", reject)
      .connect({
        host: opts.host,
        port: 22,
        username: opts.user,
        password: opts.password,
        readyTimeout: 30000,
      });
  });

  console.log(`Connected to ${opts.host}`);

  const sftp = await new Promise((resolve, reject) => {
    conn.sftp((err, s) => (err ? reject(err) : resolve(s)));
  });

  await mkdir(sftp, appDir);
  await mkdir(sftp, `${appDir}/deploy`);

  for (const [rel, remoteRel] of files) {
    const local = path.join(opts.smsDir, rel);
    const remote = `${appDir}/${remoteRel}`;
    console.log(`Upload ${rel} -> ${remote}`);
    await put(sftp, local, remote);
  }

  await exec(
    conn,
    `sed -i 's/\\r$//' ${appDir}/deploy/install-remote.sh && chmod 600 ${appDir}/.env && chmod +x ${appDir}/deploy/install-remote.sh && SMS_DOMAIN=${opts.domain} CERTBOT_EMAIL=crew.matchplay@gmail.com bash ${appDir}/deploy/install-remote.sh`
  );

  conn.end();
  console.log("\nDeploy complete.");
}

main().catch((err) => {
  console.error(err.message || err);
  process.exit(1);
});

File: crewplay-platform/scripts/deploy-sms-vps.py
#!/usr/bin/env python3
"""Deploy sms-proxy to AlmaLinux VPS via SSH/SFTP."""
from __future__ import annotations

import argparse
import os
import stat
import subprocess
import sys
import textwrap


def ensure_paramiko():
    try:
        import paramiko  # noqa: F401
        return
    except ImportError:
        subprocess.check_call([sys.executable, "-m", "pip", "install", "paramiko", "-q"])


def main() -> int:
    ensure_paramiko()
    import paramiko

    p = argparse.ArgumentParser()
    p.add_argument("--host", required=True)
    p.add_argument("--user", default="root")
    p.add_argument("--password", required=True)
    p.add_argument("--domain", default="sms.crewplay.tw")
    p.add_argument("--sms-dir", required=True)
    args = p.parse_args()

    app_dir = "/opt/crewplay-sms-proxy"
    sms_dir = os.path.abspath(args.sms_dir)
    env_path = os.path.join(sms_dir, ".env")
    if not os.path.isfile(env_path):
        print(f"Missing {env_path}", file=sys.stderr)
        return 1

    files = [
        ("server.mjs", "server.mjs"),
        (".env", ".env"),
        ("deploy/crewplay-sms-proxy.service", "deploy/crewplay-sms-proxy.service"),
        ("deploy/nginx-sms.conf.example", "deploy/nginx-sms.conf.example"),
        ("deploy/install-remote.sh", "deploy/install-remote.sh"),
    ]

    client = paramiko.SSHClient()
    client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
    print(f"Connecting to {args.user}@{args.host}...")
    client.connect(args.host, username=args.user, password=args.password, timeout=30)

    sftp = client.open_sftp()

    def mkdir_p(remote: str) -> None:
        parts = remote.strip("/").split("/")
        cur = ""
        for part in parts:
            cur += "/" + part
            try:
                sftp.mkdir(cur)
            except OSError:
                pass

    mkdir_p(app_dir)
    mkdir_p(f"{app_dir}/deploy")

    for local_rel, remote_rel in files:
        local = os.path.join(sms_dir, local_rel.replace("/", os.sep))
        remote = f"{app_dir}/{remote_rel}"
        print(f"Upload {local_rel} -> {remote}")
        sftp.put(local, remote)
        if local_rel == ".env" or local_rel.endswith(".sh"):
            sftp.chmod(remote, stat.S_IRUSR | stat.S_IWUSR)

    sftp.chmod(f"{app_dir}/deploy/install-remote.sh", stat.S_IRWXU)

    cmd = textwrap.dedent(
        f"""
        set -e
        export SMS_DOMAIN={args.domain}
        export CERTBOT_EMAIL=crew.matchplay@gmail.com
        bash {app_dir}/deploy/install-remote.sh
        """
    ).strip()
    print("Running remote install...")
    stdin, stdout, stderr = client.exec_command(cmd, get_pty=True, timeout=900)
    out = stdout.read().decode("utf-8", errors="replace")
    err = stderr.read().decode("utf-8", errors="replace")
    code = stdout.channel.recv_exit_status()
    print(out)
    if err.strip():
        print(err, file=sys.stderr)
    client.close()
    if code != 0:
        print(f"Remote install failed (exit {code})", file=sys.stderr)
        return code
    print("Deploy complete.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
